src/io.rs

```diff
@@ -110,7 +147,86 @@ pub async fn prepare_directories() {
         "./results/base",
         "./results/native",
         "./results/dependencies",
+        "./results/class_hashes",
     ];
 
     future::join_all(paths.iter().map(tokio::fs::create_dir_all)).await;
 }
+
+pub fn try_deserialize<T, P>(path: &P) -> Result<T, anyhow::Error>
+where
+    T: for<'a> Deserialize<'a>,
+    P: AsRef<Path>,
+{
+    let path_str = path.as_ref().to_str().unwrap_or("<Failed to unwrap path.>");
+
+    if !path.as_ref().exists() {
+        return Err(anyhow!("Path does not exist: {}", path_str));
+    }
+    let file = OpenOptions::new()
+        .read(true)
+        .open(path)
+        .context(format!("Opening file: {}", path_str))?;
+    serde_json::from_reader(file).context(format!("Deserializing from file: {}", path_str))
+}
+
+/// Attempts to deserialize from `path` and return the result.
+/// If deserialization fails, returns the result of `f`.
+///
+/// If `f` is a closure that returns `T`, use `deserialize_or`.
+pub fn try_deserialize_or<T, F, P>(path: &P, f: F) -> Result<T, anyhow::Error>
+where
+    T: for<'a> Deserialize<'a>,
+    F: Fn() -> Result<T, anyhow::Error>,
+    P: AsRef<Path>,
+{
+    match try_deserialize(path) {
+        Ok(obj) => Ok(obj),
+        Err(err) => {
+            debug!("Failed to deserialize: {err}");
+            f()
+        }
+    }
+}
+
+/// Attempts to deserialize from `path` and return the result.
+/// If deserialization fails, returns the result of `f`.
+///
+/// If `f` is a closure that returns `Result<T, anyhow::Error>`, use `try_deserialize_or`.
+pub fn deserialize_or<T, F, P>(path: &P, f: F) -> T
```

Comment:
No need for path to &P, it already can be just P since you are using the `AsRef` which allows to either work with the owned type (`PathBuf`) or a reference to it (`Path`).

This applies to `try_deserialize` and `try_deserialize_or` as well

Reply:
Changed.